src/core/Akka/Actor/CoordinatedShutdown.cs

```diff
@@ -0,0 +1,634 @@
+
+        private volatile bool _runningClrHook = false;
+
+        /// <summary>
+        /// INTERNAL API
+        /// 
+        /// Signals when CLR shutdown hooks have been completed
+        /// </summary>
+        internal Task<Done> ClrShutdownTask => _hooksRunPromise.Task;
+
+        /// <summary>
+        /// Add a task to a phase. It doesn't remove previously added tasks.
+        /// 
+        /// Tasks added to the same phase are executed in parallel without any
+        /// ordering assumptions. Next phase will not start until all tasks of
+        /// previous phase have completed.
+        /// </summary>
+        /// <param name="phase">The phase to add this task to.</param>
+        /// <param name="taskName">The name of the task to add to this phase.</param>
+        /// <param name="task">The delegate that produces a <see cref="Task"/> that will be executed.</param>
+        /// <remarks>
+        /// Tasks should typically be registered as early as possible after system
+        /// startup. When running the <see cref="CoordinatedShutdown"/> tasks that have been
+        /// registered will be performed but tasks that are added too late will not be run.
+        /// 
+        /// 
+        /// It is possible to add a task to a later phase from within a task in an earlier phase
+        /// and it will be performed.
+        /// </remarks>
+        public void AddTask(string phase, string taskName, Func<Task<Done>> task)
+        {
+            if (!_knownPhases.Contains(phase))
+                throw new ConfigurationException($"Unknown phase [{phase}], known phases [{string.Join(",", _knownPhases)}]. " +
+                    "All phases (along with their optional dependencies) must be defined in configuration.");
+
+            ImmutableList<Tuple<string, Func<Task<Done>>>> current;
+            if (!_tasks.TryGetValue(phase, out current))
+            {
+                if (!_tasks.TryAdd(phase,
+                    ImmutableList<Tuple<string, Func<Task<Done>>>>.Empty.Add(Tuple.Create(taskName, task))))
+                {
+                    AddTask(phase, taskName, task); // CAS failed, retry
+                }
+            }
+            else
+            {
+                if (!_tasks.TryUpdate(phase, current.Add(Tuple.Create(taskName, task)), current))
+                {
+                    AddTask(phase, taskName, task); // CAS failed, retry
+                }
+            }
+        }
+
+        /// <summary>
+        /// Add a shutdown hook that will execute when the CLR process begins
+        /// its shutdown sequence, invoked via <see cref="AppDomain.ProcessExit"/>.
+        /// 
+        /// Added hooks may run in any order concurrently, but they are run before
+        /// the Akka.NET internal shutdown hooks execute.
+        /// </summary>
+        /// <param name="hook">A task that will be executed during shutdown.</param>
+        internal void AddClrShutdownHook(Func<Task<Done>> hook)
```

Comment:
There was a discussion about this on Gitter... It seemed extraneous, to some extent, since the notion of a JvmShutdownHook is actually built into the Java VM itself, whereas I invented our own version here that doesn't necessarily have the same level of support out of the box. I thought it best to leave it as `internal` for now, see if there's demand for it / how reliable it is.